mediapipe/graphs/instantmotiontracking/calculators/matrices_manager.cc

```diff
@@ -0,0 +1,221 @@
+          .Get<std::vector<Anchor>>();
+
+  // Device IMU Data definitions
+  const float roll = cc->Inputs().Tag(kIMUDataTag).Get<float[]>()[0];
+  const float pitch = cc->Inputs().Tag(kIMUDataTag).Get<float[]>()[1];
+  const float yaw = cc->Inputs().Tag(kIMUDataTag).Get<float[]>()[2];
+
+  for (Anchor anchor : translation_data) {
+    int id = anchor.sticker_id;
+
+    TimedModelMatrixProto* model_matrix =
+        model_matrix_list->add_model_matrix();
+    model_matrix->set_id(id);
+
+    float rotation = getUserRotation(user_rotation_data, id);
+    float scaling = getUserScaling(user_scaling_data, id);
+
+    Eigen::Vector3f translation_vector = generateTranslationVector(anchor, scaling);
+    Eigen::Matrix3f rotation_submatrix = generateRotationSubmatrix(roll, pitch, yaw, rotation);
+
+    Matrix4fRM mvp_matrix = generateEigenModelMatrix(translation_vector, rotation_submatrix);
+
+    // The generated model matrix must be mapped to TimedModelMatrixProto (col-wise)
+    for (int i = 0; i < mvp_matrix.rows(); ++i) {
+      for (int j = 0; j < mvp_matrix.cols(); ++j) {
+        model_matrix->add_matrix_entries(mvp_matrix(j, i));
+      }
+    }
+  }
+
+  cc->Outputs()
+      .Tag(kModelMatricesTag)
+      .Add(model_matrices.release(), cc->InputTimestamp());
+
+  return ::mediapipe::OkStatus();
+}
+
+// Using the tracked translation data, generate a vector for MVP translation
+Eigen::Vector3f MatricesManagerCalculator::generateTranslationVector(Anchor tracked_anchor, float user_scaling_increment) {
+  // Convert from normalized [0.0-1.0] to openGL on-screen coordinates
+  float z = (initial_z_ + user_scaling_increment) * tracked_anchor.z;
+
+  float y_minimum = z * (tan(vertical_fov_radians_ /2));
+  // Minimum y value appearing on screen at z distance
+  float x_minimum = y_minimum * (1.0/aspect_ratio_);
+  // Minimum x value appearing on screen at z distance
+
+  float x = (tracked_anchor.x * (-x_minimum - x_minimum)) + x_minimum;
+  float y = (tracked_anchor.y * (-y_minimum - y_minimum)) + y_minimum;
+
+  Eigen::Vector3f t_vector(x,y,z);
+  return t_vector;
+}
+
+// Generate the submatrix defining rotation using IMU data and user rotations
+Eigen::Matrix3f MatricesManagerCalculator::generateRotationSubmatrix(float roll, float pitch, float yaw, float user_rotation_radians) {
+  Eigen::Matrix3f r_submatrix;
+  r_submatrix =
+      Eigen::AngleAxisf(yaw - user_rotation_radians, Eigen::Vector3f::UnitY()) *
+      Eigen::AngleAxisf(pitch, Eigen::Vector3f::UnitZ()) *
+      Eigen::AngleAxisf(roll - (M_PI / 2), Eigen::Vector3f::UnitX());
+  return r_submatrix;
+}
+
+// Generates a model matrix via Eigen with appropriate transformations
+Matrix4fRM MatricesManagerCalculator::generateEigenModelMatrix(
+    Eigen::Vector3f translation_vector, Eigen::Matrix3f rotation_submatrix) {
+  // Define basic empty model matrix
+  Matrix4fRM mvp_matrix;
+  mvp_matrix << 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0,
```

Comment:
super-nit: no need to pipe all values here if we just want to set the very last one to 1.0